include/sys/sys_heap.h

```diff
@@ -142,6 +142,26 @@ void sys_heap_free(struct sys_heap *h, void *mem);
  */
 void *sys_heap_realloc(struct sys_heap *heap, void *ptr, size_t bytes);
 
+/** @brief Expand the size of an existing allocation with alignment
+ *
+ * Behaves like sys_heap_realloc(), except that the returned memory
+ * will have a starting address in memory which is a multiple of the
+ * specified power-of-two alignment value in bytes.
+ *
+ * The alignment is guaranteed if ptr has same or stronger alignment
+ * than requested by the align parameter. That is, the typical use case
+ * is to call sys_heap_aligned_alloc() and sys_heap_aligned_realloc()
+ * with the same align argument.
+ *
+ * @param heap Heap from which to allocate
+ * @param ptr Original pointer returned from a previous allocation
+ * @param align Alignment in bytes, must be a power of two
+ * @param bytes Number of bytes requested for the new block
+ * @return Pointer to memory the caller can now use, or NULL
+ */
+void *sys_heap_aligned_realloc(struct sys_heap *heap, void *ptr, size_t align,
+			       size_t bytes);
```

Comment:
Not opposed to the API, exactly, but this mechanism being introduced seems senselessly complicated.  The heap code already guarantees an 8-byte alignment, which is AFAIK sufficient for all data member alignment requirements on all of our platforms.

Can't you just put some BUILD_ASSERT() statements in and call it a day?  Why are we mucking in the heap code to add support for a feature we already have?

Reply:
The test case in the PR fails unless the implementation is changed.

For example, if the top commit (`malloc.c`) in the PR is reverted, then the test reports that `malloc()` gives 4-byte alignment on riscv32, xtensa and sparc:
```
    Assertion failed at ZEPHYR_BASE/tests/lib/mem_alloc/src/main.c:140: test_malloc_align: (aligned_to_all_builtin_types(aptr) is false)
0x8000159c: pointer not suitable for built-in types
```

I made an attempt to define `z_malloc_heap_mem` in terms of `uint64_t` (instead of `char`) and verified the aligment in `zephyr.map`. Unfortunately that did not help and `malloc()` still gave 4-byte aligned pointers at run-time.

This is unexpected if the heap is guaranteeing 8-byte alignment on 32-bit ABI. Could something else be wrong in how `malloc.c` uses the heap code, which can clarify this behavior?

A related topic is the question of what kind of alignment we want to provide in the "minimal" implementation of `malloc()` and `realloc()`. The theoretical answer is "to any type of object". A more practical approach would be to make sure we cover at least `alignof()` of `double`, `uint64_t`, `void *` (but skip `long double` and the `complex.h` types).

Reply:
After investigating this a bit more I made the following observation of the run-time behavior (not diving into the code):

If `CONFIG_MINIMAL_LIBC_MALLOC_ARENA_SIZE` is `262151` or less, then `sys_heap_alloc()` sometimes returns 4-byte aligned pointers. But if that config is `262152` (256 KiB) or greater, then it appears to always give 8-byte aligned pointers. (The `mem_alloc` test in this PR uses a 2048 byte heap.)

There is a comment at the top of `heap.h` indicating that a different approach is taken by the implementation depending on heap size (256 KiB or larger, vs less). https://github.com/zephyrproject-rtos/zephyr/blob/master/lib/os/heap.h#L27

Reply:
I'm still not following.  Types like doubles on all of our 32 bit platforms currently are perfectly safe to align on 4-byte boundaries.  Is this an argument about runtime behavior or standards compliance or what?

If there's a real concern that a specific platform (which one?!) simply can never allow a 4-byte aligned heap block, then the fix should be to use the "big" heap mode, which is precisely what it's for.

Reply:
The 32-bit SPARC architecture specification requires that the CPU implementation trap any 64-bit load/store which is not naturally aligned. This includes access to objects of the C types `double` and `uint64_t`. The 8-byte alignment requirement applies for example to the 32-bit in-tree Zephyr board `qemu_leon3` which is run as part of CI. Issue is described in #30723.

I have followed your advice and removed the heap and malloc changes from the PR. The test case for data access is still in. Thank you for your feedback.

Moving forward, using the "big" heap mode sounds like a really good solution. How can it be enabled?